### PR TITLE
Feature/user register client

### PR DESCRIPTION
**Pull request by pradtke**
Allow users with a given entitlement the ability to register their own clients.

### CODE REVIEW

File: lib/Controller/ClientEditController.php [hunk 1]
```diff
@@ -91,7 +94,9 @@ public function __invoke(ServerRequest $request)
                 $data['scopes'],
                 (bool) $data['is_enabled'],
                 (bool) $data['is_confidential'],
-                $data['auth_source']
+                $data['auth_source'],
+                //TODO: can this be edited, and if so by the owner or just the ssp admin?
+                $data['owner'] = $this->authContextService->isSspAdmin() ? $data['owner'] : $client->getOwner()
```
Comment (sgomez) [inline lib/Controller/ClientEditController.php:99]: IMHO owner is not a mutable field. If a RP has a new admin, he/she can register a new client and avoid to bother admins.

Comment (pradtke) [reply lib/Controller/ClientEditController.php]: I've removed the ability to edit owner.

**Comment by pradtke** [comment]
I think this is ready to merge.

File: README.md [hunk 2]
```diff
@@ -194,6 +194,22 @@ Once the database schema has been created, you can open the _Federation_ tab fro
 
 The module lets you create, read, update and delete all the RP you want. To see the client id and the client secret press the show button.
 
+### Client self registration
+
+You can allow users to register their own clients.
```
Comment (cicnavi) [inline README.md:199]: Maybe also add description on how to gracefully disable this feature. Is it best to leave out 'permissions' config or leave ['permissions']['client'] empty?

File: config-templates/module_oidc.php [hunk 3]
```diff
@@ -43,6 +43,12 @@
 
     // useridattr is the attribute-name that contains the userid as returned from idp
     'useridattr' => 'uid',
+    'permissions' => [
```
Comment (cicnavi) [inline config-templates/module_oidc.php:46]: I would add a short explanation about this config option, like 'option to allow users to register their own clients...', and also short indication on how to gracefully disable this feature.

File: hooks/hook_frontpage.php [hunk 4]
```diff
@@ -43,7 +43,7 @@ function oidc_hook_frontpage(&$links)
     }
 
     $links['federation']['oidcregistry'] = [
-        'href' => \SimpleSAML\Module::getModuleURL('oidc/clients/'),
+        'href' => \SimpleSAML\Module::getModuleURL('oidc/admin-clients/'),
```
Comment (cicnavi) [inline hooks/hook_frontpage.php:46]: There is also hooks/hook_federationpage.php (used when usenewui option is enabled).

**Review by cicnavi — COMMENTED** [review]
Really great contribution! All I would like to know is what would be the best way to disable it so that implementer that won't use it can be sure on how to do it.

File: hooks/hook_frontpage.php (hunk 4 above)
Comment (pradtke) [reply hooks/hook_frontpage.php:46]: fixed

File: config-templates/module_oidc.php (hunk 3 above)
Comment (pradtke) [reply config-templates/module_oidc.php:55]: Make sense.

File: README.md (hunk 2 above)
Comment (pradtke) [reply README.md:199]: Make sense

**Comment by pradtke** [comment]
@cicnavi I have added information on how to disable, and fixed the other link issue.

**Review by cicnavi — APPROVED** [review]

**Pull request merged**